Source/core/FileObserver.h

```diff
@@ -245,42 +249,173 @@ class FileSystemMonitor : public Core::IResource {
     };
 
 private:
-    class Worker : public Core::Thread {
+    class Context {
+    private:
+        using ClientList = std::list<ICallback*>;
     public:
-        Worker() = delete;
-        Worker(const Worker&) = delete;
-        Worker& operator= (const Worker&) = delete;
-        Worker(FileSystemMonitor& parent) 
-            : _parent(parent) {
-            _syncEvent = ::CreateEvent(nullptr, true, false, nullptr);
+        Context() = delete;
+        Context(const Context&) = delete;
+        Context& operator= (const Context&) = delete;
+
+        Context(const string& pathName)
+            : _adminLock()
+            , _filename(pathName)
+            , _file(INVALID_HANDLE_VALUE)
+            , _clients() {
+            _file = CreateFile(pathName.c_str(),
+                FILE_LIST_DIRECTORY,
+                FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                NULL,
+                OPEN_EXISTING,
+                FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED,
+                NULL);
+
+            if (_file != INVALID_HANDLE_VALUE) {        
+                _overlapped.hEvent = CreateEvent(NULL, FALSE, 0, NULL);
+
+                if (ReadDirectoryChangesW(
+                    _file, _buffer, sizeof(_buffer), FALSE,
+                    FILE_NOTIFY_CHANGE_FILE_NAME |
+                    FILE_NOTIFY_CHANGE_DIR_NAME,
+                    NULL, &_overlapped, NULL) == FALSE) {
+                    TRACE_L1("Could not start observing: %s", _filename.c_str());
+                }
+            }
         }
-        ~Worker() override {
+        ~Context() {
+            if (_file != INVALID_HANDLE_VALUE) {
+                ::CloseHandle(_file);
+                if (_overlapped.hEvent != INVALID_HANDLE_VALUE) {
+                    ::CloseHandle(_overlapped.hEvent);
+                }
+            }
         }
 
     public:
-        HANDLE Event() {
-            return(_syncEvent);
+        void Register(ICallback* client) {
+            ASSERT(std::find(_clients.begin(), _clients.end(), client) == _clients.end());
+
+            _clients.push_back(client);
```

Comment:
Not protected I guess for a reason? (performance and not used multithreaded)

Reply:
It is protected by the outer class. Inner class can not be directly accessed. So not required to lock (unless we want to double lock :-) )